### PR TITLE
Implement custom JUL bridge

### LINKED ISSUE #94613: Configure Java Util Logging bridge
Elasticsearch uses log4j as its logger. java.util.logging may be used by some dependencies (eg Lucene now uses it in some places). Currently these log messages are captured by our stdout/stderr -> log4j because java.util.logging has a default implementation which writes to stdout/stderr.

log4j includes a bridge jar that configures java.util.logging to go directly to log4j logging. Using that bridge is superior to capturing the same log messages on stdout/stderr because it provides better control over muting classes or packages log messages. We should consider adding the java.util.logging bridge.

### PR DESCRIPTION
**Pull request by rjernst**
The log4j JUL bridge turned out to have issues because it relied on java beans. This commit implements a custom bridge between JUL and Log4j.

closes #94613

### CODE REVIEW

**Review by uschindler — COMMENTED** [review]
Hahah much more easy to understand. That's a quite simple imp and we need no external dependency.

Once we move to own logging framework this can easily be ported to directly use ESLogger or similar.

**Comment by uschindler** [comment]
Just the question: How did the beans shit break ES?

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+    );
+
+    private static final TreeMap<Integer, Level> sortedLevelMap = new TreeMap<>(
+        levelMap.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().intValue(), Map.Entry::getValue))
+    );
+
+    public static void install() {
+        var rootJulLogger = java.util.logging.LogManager.getLogManager().getLogger("");
+        // clear out any other handlers, so eg we don't also print to stdout
+        for (var existingHandler : rootJulLogger.getHandlers()) {
```
Comment (uschindler) [inline server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:50]: is it a good idea to remove elements while executing the loop?

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+        Level.INFO,
+        java.util.logging.Level.FINE,
+        Level.DEBUG,
+        java.util.logging.Level.FINEST,
+        Level.TRACE,
+        java.util.logging.Level.ALL,
+        Level.ALL
+    );
+
+    private static final TreeMap<Integer, Level> sortedLevelMap = new TreeMap<>(
```
Comment (uschindler) [inline server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:43]: You could also write:
```java
 sortedLevelMap = levelMap.entrySet().stream().collect(Collectors.toMap(e -> e.getKey().intValue(), Map.Entry::getValue, (k1, k2) -> throw new UnsupportedOperationException(), TreeMap::new));
```

(I hate that theres no Collector method that allows to specify custom map factory, but uses default merge function. 🤮

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java (hunk 1 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:50]: It's an array, so it doesn't actually change as remove is called.

**Comment by rjernst** [comment]
> How did the beans shit break ES?

PropertyChangeListener was unavailable at runtime because we don't read the java.desktop module.

**Comment by rjernst** [comment]
@elasticsearchmachine run elasticsearch-ci/part-2

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java (hunk 2 above)
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/logging/JULBridge.java]: Thanks! I used the Maps method and changed to NavigableMap

**Comment by rjernst** [comment]
I addressed the feedback, and switched this to use the ES logging api instead of log4j directly.

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+import java.util.Map;
+import java.util.NavigableMap;
+import java.util.TreeMap;
+import java.util.logging.Handler;
+import java.util.logging.LogRecord;
+import java.util.stream.Collectors;
+
+/**
+ * A Java Util Logging handler that writes log messages to log4j.
```
Comment (uschindler) [inline server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:24]: This should be EsLogger instead log4j

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java (hunk 1 above)
Comment (pgomulka) [reply server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:50]: should be safe. the getHandlers return a copied array of handlers

**Review by pgomulka — COMMENTED** [review]
It looks good, I am worried that we perhaps are incorrectly formatting parameterised messages? or have I missed something?
I made some suggestions in https://github.com/rjernst/elasticsearch/pull/4

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+    @Override
+    public void publish(LogRecord record) {
+        Logger logger = LogManager.getLogger(record.getLoggerName());
+        Level level = translateJulLevel(record.getLevel());
+        Throwable thrown = record.getThrown();
+        if (thrown == null) {
+            logger.log(level, record.getMessage());
+        } else {
+            logger.log(level, record::getMessage, thrown);
```
[the remark this answers is not in the document]
Comment (rjernst) [reply server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:66]: I don't think we do? AFAICT the formatting for JUL is just the string format of the line, not substitutions. Log4j handles that for us, all we need to do is pass the message intact.

**Comment by rjernst** [comment]
@pgomulka You are right, I added handling for null and formatted messages, see https://github.com/elastic/elasticsearch/pull/96872/commits/e8900c0e5615f65f695879ec5ac3c528607c1599

File: server/src/main/java/org/elasticsearch/common/logging/JULBridge.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+        if (thrown == null) {
+            logger.log(level, message);
+        } else {
+            logger.log(level, () -> message, thrown);
+        }
+    }
+
+    private Level translateJulLevel(java.util.logging.Level julLevel) {
+        Level log4jLevel = levelMap.get(julLevel);
```
Comment (uschindler) [inline server/src/main/java/org/elasticsearch/common/logging/JULBridge.java:81]: Still log4j 🤨💪

File: server/src/test/java/org/elasticsearch/common/logging/JULBridgeTests.java [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.Level;
```
Comment (pgomulka) [inline server/src/test/java/org/elasticsearch/common/logging/JULBridgeTests.java:11]: I guess this is fine, but perhaps we should be making sure that the bridge works between jul-es_logging_api

**Review by pgomulka — APPROVED** [review]
LGTM, 
we assert about log4j-api in testing, which I think is ok. Making this working with ES api would probably require changes to the test framework.

**Comment by rjernst** [comment]
I removed log4j from the impl, but it remains in tests because that is what the mock logger works on. We can followup with a way for the ES logger to be mocked.

**Pull request merged**